hack/tools/janitor/janitor.go

```diff
@@ -162,12 +204,152 @@ func (s *janitor) deleteVSphereVMs(ctx context.Context, folder string) error {
-	for _, t := range tasks {
-		if err := t.Wait(ctx); !ignoreErrors && err != nil {
+// deleteObjectChildren deletes all child objects in a given object in vSphere if they don't
+// contain any virtual machine.
+// An object only gets deleted if:
+// * it does not have any children of a different type
+// * the timestamp field's value is before s.maxCreationDate
+// If an object does not yet have a field, the janitor will add the field to it with the current timestamp as value.
+func (s *janitor) deleteObjectChildren(ctx context.Context, inventoryPath string, objectType string) error {
+	if !slices.Contains([]string{"ResourcePool", "Folder"}, objectType) {
+		return fmt.Errorf("deleteObjectChildren is not implemented for objectType %s", objectType)
+	}
+
+	if inventoryPath == "" {
+		return fmt.Errorf("cannot use empty string to delete children of type %s", objectType)
+	}
+
+	log := ctrl.LoggerFrom(ctx).WithName(fmt.Sprintf("%sChildren", objectType)).WithValues(objectType, inventoryPath)
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	log.Info("Deleting empty children")
+
+	// Recursively list all objects of the given objectType below the inventoryPath.
+	managedEntities, err := recursiveList(ctx, inventoryPath, s.vSphereClients.Govmomi, s.vSphereClients.Finder, s.vSphereClients.ViewManager, objectType)
+	if err != nil {
+		return err
+	}
+
+	// Build a map which notes if an object has children of a different type.
+	// Later on we will use that information to not delete objects which have children.
+	hasChildren := map[string]bool{}
+	for _, e := range managedEntities {
+		// Check if the object has children, because we only want to delete objects which have children of a different type.
+		children, err := recursiveList(ctx, e.element.Path, s.vSphereClients.Govmomi, s.vSphereClients.Finder, s.vSphereClients.ViewManager)
+		if err != nil {
+			return err
+		}
+		// Mark e to have children, if there are children which are of a different type.
+		for _, child := range children {
+			if child.entity.Reference().Type == objectType {
+				continue
+			}
+			hasChildren[e.element.Path] = true
+			break
+		}
+	}
+
+	// Get key for the deletion marker.
+	deletionMarkerKey, err := s.vSphereClients.FieldsManager.FindKey(ctx, vSphereDeletionMarkerName)
+	if err != nil && err != object.ErrKeyNameNotFound {
+		return errors.Wrapf(err, "finding custom field %q", vSphereDeletionMarkerName)
+	}
+
+	// Only create the deletion marker if we are not on dryRun.
+	if deletionMarkerKey == -1 {
+		log.Info("Creating the deletion field")
+
+		if !s.dryRun {
+			_, err := s.vSphereClients.FieldsManager.Add(ctx, vSphereDeletionMarkerName, "ManagedEntity", nil, nil)
+			if err != nil {
+				return errors.Wrapf(err, "creating custom field %q", vSphereDeletionMarkerName)
+			}
+
+			deletionMarkerKey, err = s.vSphereClients.FieldsManager.FindKey(ctx, vSphereDeletionMarkerName)
+			if err != nil && err != object.ErrKeyNameNotFound {
```

Comment:
Why is it okay to get the ErrKeyNameNotFound error here?

Reply:
not okay :-) refactored this part.